Errors on frontend

jlurgo opened the pull request:
Frist pass for showing errors on the frontend:
(Video made showing INFO level logs)
[Screencast from 2025-06-25 14-43-52.webm](https://github.com/user-attachments/assets/64233a7d-6ef3-453f-ab80-b997996646a1)

jlurgo commented:
I saved the logs in MqttSync using the timestamp as the dictionary key, IDK if it's the best way of doing it:
- advantages:
  - adding new ones is easy and lightweight, just a simple update
  - you mentioned that arrays in mqttsync are tricky
- problems:
  - timestamp may be repeated but it's easy to workaround it
  - to dismiss error logs I had to use MqttSync.clear that takes a while and has some usage warnings in the code

chfritz commented:
> I saved the logs in MqttSync using the timestamp as the dictionary key

Never use numbers as keys. This is actually part of the issue with Arrays in MQTTSync. Keys that are numbers are interpreted as array indices, so if you publish `/some/topic/1234 = "myvalue"` then it will create
```js
{ 
  some: {
    topic: [ 1233 empty fields,  "myvalue" ] 
  }
}
```
Creating the large arrays (even if mostly empty) is inefficient and I think MQTTSync doesn't handle them well at all, it might even fill all the empty fields with `null`, not sure. This is also why we prefix device names with `d_`, in case the user uses a `TR_INSTALL_HASH` that is a number.

chfritz commented:
Please double check in DataCache that UUIDs cannot be misinterpreted as numbers if they start with a digit. Might be easyer (and shorter) to use [getdatebase52](https://transitiverobotics.com/docs/sdk/common/#getdatebase52) instead.

For clearing we don't need to use the clear function. Since we've already received the value we can just set the topic to `null` (and publish it). That way it will be gone in React right away.

We'll also need to add some auto-cleanup of these log lines on the server to not overwhelm the user with errors that are many days old. Also, the current design is very "error" focused (topic name and red badge), but depending on how it is configured these will not all be errors. For the visualization in the modal, I would not change the display, i.e., not break it up into bullet points, but just show it as text, just like we have it with the "get log".

jlurgo commented:
> We'll also need to add some auto-cleanup of these log lines on the server to not overwhelm the user with errors that are many days old. Also, the current design is very "error" focused (topic name and red badge), but depending on how it is configured these will not all be errors. For the visualization in the modal, I would not change the display, i.e., not break it up into bullet points, but just show it as text, just like we have it with the "get log".

Yes, on the front end I just focused on errors (red badges for error count) that's what we want to raise awareness for users, right?
If we want to show other log levels I'd use a different UI for them (I'm using INFO level, but only for testing purposes) 

> For clearing we don't need to use the clear function. Since we've already received the value we can just set the topic to `null` (and publish it). That way it will be gone in React right away.

That's what I tested first: 
on component load:` mqttSync.publish(`${prefix}/+/errorLogs/#`);`
then on clear button click:` mqttSync.data.update(agentErrorLogsTopic, null); `
Data is cleared immediately on the frontend but logs come back after refreshing the page.

jlurgo commented:
OK that's what I thought, so I should iterate over all logs and set them to null? I though **mqttSync.clear** was doing something similar... maybe not quite that, since it subscribes to retained messages on the mqtt topic to clear them? Should I iterate?

chfritz commented:
getRandomId uses base36, i.e., `[0-9a-z]`, so it will still generate values that start with digits. I specifically added `getdatebase52` because I had the exact same issue in the video-recorder (needing values I can use as topic keys), which is why it is using only `[a-zA-Z]`.

jlurgo commented:
Ok, now I'm storing the error count in mqttSync, clearing the count when we truncate the log file on log rotation.
There's still some work to do on the styling side.

I'll investigate querying hyperdx/clickhouse, but I have a question:
- we agreed on it being the next step, the idea here then is to replace the rpc command that gets the logs from the agent log files with getting logs from the database?

![image](https://github.com/user-attachments/assets/85a4eb55-fc1a-480e-bc5e-c1592f31777a)

jlurgo commented:
I implemented showing live feed of log lines while looking at get logs modal dialog. 
Errors count is shown as a superindex on top of the get log buttons.
I also added the button to see logs for robot agent.

Pending:
- Improve styling for the errors counter
- Add tooltip with last error
- I've been testing this by enabling log rotation on agent restart (as it was originally) so to have the errors count to match the error logs included in the file... but I didn't push it here because we have the plan to make rotation to be based on log lines count.
- Would be great to show errors in red or any other way of highlighting errors in the log so to help visually match error count with the logs
- Maybe we can also show (and allow users to change) log level in the modal logs dialog.

[Screencast from 2025-06-30 17-49-57.webm](https://github.com/user-attachments/assets/903b1477-4935-430a-8ac8-5a27552b730d)

jlurgo commented:
I restored Log rotation on startup for now, until we work on a file size/line count based rotation, which (if I got it right) we're leaving for later

jlurgo commented:
I noted that we're always showing "stderr is empty" after the past-non-live logs.
![image](https://github.com/user-attachments/assets/e64c6a69-c063-4c54-9875-204368691971)

I tested logging errors and throwing unhandled exceptions within a capability to see if I can get some data there, but it's always empty, errors are coming within stdout data. Same happens with agent log.
In agent logs we're tee-ing stdout + stderr into the same file
For capabilities we spawn a child process, redirecting stdout and stderr to the log file
For both when we run getPkgLog we:
![image](https://github.com/user-attachments/assets/72885445-9fce-4362-a8e9-975e8f2a553e)

Maybe we can get rid of the stderr section there...

chfritz commented:
Good find. Yes, that stderr output doesn't make sense, because it's actually the stderr output of the `cat` command we use to get the log from disk, not the stderr of the process we care about.

jlurgo commented:
OH, just saw your last review, then I pushed 2 commits, I'll address it now...

I got in a small rabbit hole here, but I already emerged and I hope you like the results, 
I found that the button to get logs in robot-agent-device-header.jsx wasn't showing live logs (and I think it wasn't working due to a different version in the RPC topic)
So I fixed and added live logs + the errors count:
![image](https://github.com/user-attachments/assets/21670129-a332-4f2c-a865-9e206acf0d37)
And then I realized there was a bunch of repeated code so I packed everything logs retrieval and counting related into the GetLogButtonWithCounter component.

jlurgo commented:
Thanks for the review @chfritz , addressed!
